### PR TITLE
add sampling of near-neighbor context sequences

### PR DESCRIPTION
**Pull request by gregcaporaso**

### CODE REVIEW

**Comment by gregcaporaso** [comment]
Two Metadata-related tests are currently failing. I'm currently thinking that (1) input Metadata should require entries for all ids undergoing selection (the context sequences in this case), and (2) the `IDSelection.Metadata` should not include any ids that were not undergoing selection but were provided in the input. I have a test for each of these in my most recent commit, and both are failing. @ebolyen, do you agree with these expectations? Validating (1) in cases like this will probably require two passes through the sequences, which is unfortunate.

File: genome_sampler/subsample_neighbors.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    if len_cluster <= samples_per_cluster:
+        return list(cluster['context_id'])
+
+    unique_locales = cluster['locale'].unique()
+    if len(unique_locales) == 1:
+        # if only one locale, select context sequences spanning the range
+        # of number of mismatches
+        sorted_cluster = cluster.sort_values(by='n_mismatches')
+        indices = map(int, np.linspace(0, len_cluster-1, samples_per_cluster))
+        return [sorted_cluster.iloc[i]['context_id'] for i in indices]
```
Comment (ebolyen) [inline genome_sampler/subsample_neighbors.py:47]: I think something like:
```python
indices = np.linspace(0, len_cluster-1, samples_per_cluster).astype(int)
return sorted_cluster['context_id'].iloc[indices].tolist()
```
would be more typical

File: genome_sampler/subsample_neighbors.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    weights = ((1 / loc.map(loc.value_counts())) / len(loc.unique()))
+    # TODO: if multiple samples are being selected from the same locale, it
+    # would be good to sample across the number of mismatches to the focal
+    # sequence, as we're doing where there is only a single locale. To do that,
+    # I think we'd need to get the count of how many of each locale we want
+    # at random (like the rarefaction algorithm) and then use the approach
+    # above (which would then generalize to one or more locales).
+    result = cluster.sample(samples_per_cluster,
+                            weights=weights,
+                            random_state=random_state)
```
Comment (ebolyen) [inline genome_sampler/subsample_neighbors.py:64]: this is missing the behavior:
 https://github.com/caporaso-lab/genome-sampler/blob/30a837b42f927d5d85c9edf01e9cbed3d28a9697/genome_sampler/sample-clusters.py#L116-L124

which I think is what the TODO is trying to get at, but I'm not sure how we would do that.

File: genome_sampler/tests/test_subsample_neighbors.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+                count_obs_c2 += 1
+            if 'c3' in obs_sampled_context_seqs:
+                count_obs_c3 += 1
+            if 'c4' in obs_sampled_context_seqs:
+                count_obs_c4 += 1
+            if 'c5' in obs_sampled_context_seqs:
+                count_obs_c5 += 1
+
+        # since c2, c3, and c5 all have locale "def" and c4 has locale "hijk",
+        # so we expect to see c4 more frequently than any of the other three
```
Comment (ebolyen) [inline genome_sampler/tests/test_subsample_neighbors.py:123]: :+1:

**Comment by gregcaporaso** [comment]
Thanks for the review @ebolyen!

**Pull request merged**